File: frontend/lib/models/machine.dart
import 'package:flutter/material.dart';

/// Enum representing the status of a machine
enum MachineStatus { available, paidFor, inUse, pendingUnload, outOfOrder }

/// Enum representing the type of machine
enum MachineType { washer, dryer }

/// Model representing a status change entry in history
class StatusHistoryEntry {
  final MachineStatus status;
  final DateTime timestamp;
  final String user;

  StatusHistoryEntry({
    required this.status,
    required this.timestamp,
    required this.user,
  });

  /// Get formatted status string
  String get formattedStatus {
    switch (status) {
      case MachineStatus.available:
        return 'Available';
      case MachineStatus.paidFor:
        return 'Paid For';
      case MachineStatus.inUse:
        return 'In Use';
      case MachineStatus.pendingUnload:
        return 'Pending Unload';
      case MachineStatus.outOfOrder:
        return 'Out of Order';
    }
  }

  factory StatusHistoryEntry.fromJson(Map<String, dynamic> json) {
    return StatusHistoryEntry(
      status: MachineStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
      ),
      timestamp: DateTime.parse(json['timestamp']),
      user: json['user'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'status': status.toString().split('.').last,
      'timestamp': timestamp.toIso8601String(),
      'user': user,
    };
  }
}

/// Model representing a Telegram message notification
class TelegramMessage {
  final String message;
  final String? messageUrl;

  TelegramMessage({required this.message, this.messageUrl});

  factory TelegramMessage.fromJson(Map<String, dynamic> json) {
    return TelegramMessage(
      message: json['message'],
      messageUrl: json['message_url'],
    );
  }

  Map<String, dynamic> toJson() {
    return {'message': message, 'message_url': messageUrl};
  }
}

/// Main model representing a laundry machine
class MachineModel {
  final String id; // Format: "55W1"-"59D12" (block + type + number)
  final int blockNumber; // Must be one of: 55, 57, 59
  MachineStatus status;
  List<StatusHistoryEntry> statusHistory;
  String? estimatedFinishTime; // ISO 8601 UTC timestamp for inUse machines
  TelegramMessage? telegramMessage;

  static const List<int> validBlocks = [55, 57, 59];

  MachineModel({
    required this.id,
    required this.blockNumber,
    required this.status,
    this.statusHistory = const [],
    this.estimatedFinishTime,
    this.telegramMessage,
  }) {
    if (!validBlocks.contains(blockNumber)) {
      throw ArgumentError(
        'Block number must be one of $validBlocks, got $blockNumber',
      );
    }
  }

  /// Get machine type from ID (e.g., '57W4' -> washer)
  MachineType get type {
    if (id.contains('W')) return MachineType.washer;
    if (id.contains('D')) return MachineType.dryer;
    throw ArgumentError('Invalid machine ID format: $id');
  }

  /// Get machine number from ID (e.g., '57W4' -> 4)
  int get number {
    final match = RegExp(r'[WD](\d+)').firstMatch(id);
    if (match != null) {
      return int.parse(match.group(1)!);
    }
    throw ArgumentError('Invalid machine ID format: $id');
  }

  /// Calculate remaining time in seconds from estimated finish time
  int? get remainingTimeSeconds {
    if (estimatedFinishTime == null) return null;
    try {
      final finishTime = DateTime.parse(estimatedFinishTime!);
      final now = DateTime.now().toUtc();
      final difference = finishTime.difference(now).inSeconds;
      return difference > 0 ? difference : 0;
    } catch (e) {
      return null;
    }
  }

  /// Get formatted remaining time as MM:SS
  String get formattedRemainingTime {
    final seconds = remainingTimeSeconds;
    if (seconds == null) return '';
    final minutes = seconds ~/ 60;
    final secs = seconds % 60;
    return '${minutes.toString().padLeft(2, '0')}:${secs.toString().padLeft(2, '0')}';
  }

  /// Get formatted status as a proper string
  String get formattedStatus {
    switch (status) {
      case MachineStatus.available:
        return 'Available';
      case MachineStatus.paidFor:
        return 'Paid For';
      case MachineStatus.inUse:
        return 'In Use';
      case MachineStatus.pendingUnload:
        return 'Pending Unload';
      case MachineStatus.outOfOrder:
        return 'Out of Order';
    }
  }

  /// Get the color for the status indicator icon
  Color get statusColor {
    switch (status) {
      case MachineStatus.available:
        return Colors.green;
      case MachineStatus.pendingUnload:
        return Colors.yellow;
      case MachineStatus.paidFor:
        return Colors.blue;
      case MachineStatus.inUse:
      case MachineStatus.outOfOrder:
        return Colors.red;
    }
  }

  /// Check if the textbox should be shown
  bool get shouldShowText {
    return status == MachineStatus.inUse || status == MachineStatus.outOfOrder;
  }

  /// Get the text to display in the textbox
  String get displayText {
    if (status == MachineStatus.inUse) {
      return formattedRemainingTime;
    } else if (status == MachineStatus.outOfOrder) {
      return 'Out Of Order';
    }
    return '';
  }

  /// Get the text color for the display text
  Color get displayTextColor {
    return status == MachineStatus.outOfOrder ? Colors.red : Colors.black;
  }

  /// Update machine status and add to history
  void updateStatus(MachineStatus newStatus, String user) {
    status = newStatus;
    statusHistory.add(
      StatusHistoryEntry(
        status: newStatus,
        timestamp: DateTime.now(),
        user: user,
      ),
    );
  }

  /// Update estimated finish time (for in-use machines)
  void updateEstimatedFinishTime(String? finishTime) {
    estimatedFinishTime = finishTime;
  }

  /// Create from JSON (for API response parsing)
  factory MachineModel.fromJson(Map<String, dynamic> json) {
    return MachineModel(
      id: json['id'],
      blockNumber: json['block_number'],
      status: MachineStatus.values.firstWhere(
        (e) => e.toString().split('.').last == json['status'],
      ),
      statusHistory:
          (json['status_history'] as List<dynamic>?)
              ?.map((e) => StatusHistoryEntry.fromJson(e))
              .toList() ??
          [],
      estimatedFinishTime: json['estimated_finish_time'],
      telegramMessage: json['telegram_message'] != null
          ? TelegramMessage.fromJson(json['telegram_message'])
          : null,
    );
  }

  /// Convert to JSON (for API requests)
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'block_number': blockNumber,
      'status': status.toString().split('.').last,
      'status_history': statusHistory.map((e) => e.toJson()).toList(),
      'estimated_finish_time': estimatedFinishTime,
      'telegram_message': telegramMessage?.toJson(),
    };
  }

  /// Copy with method for immutable updates
  MachineModel copyWith({
    String? id,
    int? blockNumber,
    MachineStatus? status,
    List<StatusHistoryEntry>? statusHistory,
    String? estimatedFinishTime,
    TelegramMessage? telegramMessage,
  }) {
    return MachineModel(
      id: id ?? this.id,
      blockNumber: blockNumber ?? this.blockNumber,
      status: status ?? this.status,
      statusHistory: statusHistory ?? this.statusHistory,
      estimatedFinishTime: estimatedFinishTime ?? this.estimatedFinishTime,
      telegramMessage: telegramMessage ?? this.telegramMessage,
    );
  }
}

File: backend/src/models/machine.py
"""Model for the laundry machine system"""
from datetime import datetime
from typing import TypedDict

from src.models.status_history import StatusHistoryEntry, StatusHistoryEntryDict
from src.models.telegram_message import TelegramMessage, TelegramMessageDict
from src.types.enums import MachineStatus, MachineType

class MachineDict(TypedDict):
    id: str
    block_number: int
    status: str
    status_history: list[StatusHistoryEntryDict]
    remaining_time_seconds: int | None
    telegram_message: TelegramMessageDict | None


class Machine:
    """Model for laundry machine"""
    _VALID_BLOCKS = {55, 57, 59}
    _MAX_RUN_TIME_SECONDS = 50 * 60  # 50 minutes in seconds
    
    def __init__(
        self,
        id: str,
        block_number: int,
        status: MachineStatus = MachineStatus.AVAILABLE,
        status_history: list[StatusHistoryEntry] | None = None,
        remaining_time_seconds: int | None = None,
        telegram_message: TelegramMessage | None = None
    ):
        if block_number not in self._VALID_BLOCKS:
            raise ValueError(f"Block number must be one of {self._VALID_BLOCKS}, got {block_number}")
        
        self._id = id
        self._block_number = block_number
        self._status = status
        self._status_history = status_history or []
        self._remaining_time_seconds = remaining_time_seconds
        self._telegram_message = telegram_message

    @property
    def status(self) -> MachineStatus:
        return self._status

    @property
    def status_history(self) -> list[StatusHistoryEntry]:
        return self._status_history

    @property
    def remaining_time_seconds(self) -> int | None:
        return self._remaining_time_seconds
    
    @remaining_time_seconds.setter
    def remaining_time_seconds(self, value: int) -> None:
        if value < 0:
            raise ValueError("Remaining time cannot be negative")
        elif value > self._MAX_RUN_TIME_SECONDS:
            raise ValueError(f"Remaining time cannot exceed {self._MAX_RUN_TIME_SECONDS} seconds")
        self._remaining_time_seconds = value

    @property
    def telegram_message(self) -> TelegramMessage | None:
        return self._telegram_message
    
    @telegram_message.setter
    def telegram_message(self, message: TelegramMessage | None) -> None:
        self._telegram_message = message
    
    @property
    def id(self) -> str:
        return self._id
    
    @property
    def block_number(self) -> int:
        """Get block number"""
        return self._block_number

    @property
    def type(self) -> MachineType:
        """Get machine type from ID (e.g., '57W4' -> WASHER)"""
        # ID format: <block><type><number> (e.g., 57W4)
        for char in self._id:
            if char in ('W', 'D'):
                return MachineType.WASHER if char == 'W' else MachineType.DRYER
        raise ValueError(f"Invalid machine ID format: {self._id}")

    @property
    def number(self) -> int:
        """Get machine number from ID (e.g., '57W4' -> 4)"""
        # Extract the number part after W or D
        for i, char in enumerate(self._id):
            if char in ('W', 'D'):
                return int(self._id[i+1:])
        raise ValueError(f"Invalid machine ID format: {self._id}")

    def update_status(self, new_status: MachineStatus, user: str):
        """Update machine status and add to history"""
        self._status = new_status
        self._status_history.append(
            StatusHistoryEntry(
                status=new_status,
                timestamp=datetime.now(),
                user=user
            )
        )

    def to_dict(self) -> MachineDict:
        """Convert to dictionary"""
        return MachineDict(
            id=self._id,
            block_number=self._block_number,
            status=self._status.value,
            status_history=[entry.to_dict() for entry in self._status_history],
            remaining_time_seconds=self._remaining_time_seconds,
            telegram_message=self._telegram_message.to_dict() if self._telegram_message else None
        )

    @classmethod
    def from_dict(cls, data: MachineDict) -> 'Machine':
        """Create from dictionary"""
        return cls(
            id=data['id'],
            block_number=data['block_number'],
            status=MachineStatus(data['status']),
            status_history=[
                StatusHistoryEntry.from_dict(entry)
                for entry in data.get('status_history', [])
            ],
            remaining_time_seconds=data.get('remaining_time_seconds'),
            telegram_message=(
                None if data['telegram_message'] is None
                else TelegramMessage.from_dict(data['telegram_message'])
            ),
        )


if __name__ == "__main__":
    pass
